src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,557 @@
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class AsyncRewriter : StateMachineRewriter
+    {
+        // PROTOTYPE(async-streams): Consider making AsyncRewriter an abstract base
+
+        /// <summary>
+        /// This rewriter rewrites an async-iterator method. See async-streams.md for design overview.
+        /// </summary>
+        private sealed class AsyncIteratorRewriter : AsyncRewriter
+        {
+            private FieldSymbol _promiseOfValueOrEndField; // this struct implements the IValueTaskSource logic
+            private FieldSymbol _promiseIsActiveField;
+            private FieldSymbol _currentField; // stores the current/yieled value
+
+            internal AsyncIteratorRewriter(
+                BoundStatement body,
+                MethodSymbol method,
+                int methodOrdinal,
+                AsyncStateMachine stateMachineType,
+                VariableSlotAllocator slotAllocatorOpt,
+                TypeCompilationState compilationState,
+                DiagnosticBag diagnostics)
+                : base(body, method, methodOrdinal, stateMachineType, slotAllocatorOpt, compilationState, diagnostics)
+            {
+                Debug.Assert(method.IteratorElementType != null);
+
+                // PROTOTYPE(async-streams): Why does AsyncRewriter have logic to ignore accessibility?
+            }
+
+            protected override void GenerateMethodImplementations()
+            {
+                // IAsyncStateMachine and constructor
+                base.GenerateMethodImplementations();
+
+                // IAsyncEnumerable
+                GenerateIAsyncEnumerableImplementation_GetAsyncEnumerator();
+
+                // IAsyncEnumerator
+                GenerateIAsyncEnumeratorImplementation_WaitForNextAsync();
+                GenerateIAsyncEnumeratorImplementation_TryGetNext();
+
+                // IValueTaskSource<bool>
+                GenerateIValueTaskSourceImplementation_GetResult();
+                GenerateIValueTaskSourceImplementation_GetStatus();
+                GenerateIValueTaskSourceImplementation_OnCompleted();
+
+                // IStrongBox<ManualResetValueTaskSourceLogic<TResult>>
+                GenerateIStrongBox_get_Value();
+
+                // IAsyncDisposable
+                GenerateIAsyncDisposable_DisposeAsync();
+            }
+
+            protected override void GenerateControlFields()
+            {
+                // the fields are initialized from entry-point method (which replaces the async-iterator method), so they need to be public
+
+                base.GenerateControlFields();
+
+                // Add a field: ManualResetValueTaskSourceLogic<bool> promiseOfValueOrEnd
+                _promiseOfValueOrEndField = F.StateMachineField(
+                    F.WellKnownType(WellKnownType.System_Threading_Tasks_ManualResetValueTaskSourceLogic_T)
+                        .Construct(F.SpecialType(SpecialType.System_Boolean)),
```

Comment:
>SpecialType.System_Boolean [](start = 49, length = 26)

Perhaps not in this PR, but we need to have tests that make sure appropriate errors are reported on all missing types used in these rewriters.

Reply:
Definitely. Added a work item in https://github.com/dotnet/roslyn/issues/24037